### ISSUE TITLE
Prevent the entire navigation stack from being rebuilt from dependency changes

### ISSUE THREAD

**Issue by b3nni97** [flutter/flutter#96044]
I have the problem that all widgets in a navigator stack are rebuilt when e.g. the MediaQuery changes (opening the keyboard).
If there are many routes in the navigator it can cause the app to start janking, even though every single build method of a single route would only take about 2-3 ms. 

**Wouldn't it be possible to teach the navigator to update routes with the "newest" dependencies only when a route becomes visible again?** 

It doesn't really make sense to rebuild the whole app on an opening keyboard when only one page is visible.

If I have overlooked something how to solve this problem I would be very grateful for suggestions.

**Comment by maheshmnj** [comment]
Hi @b3nni97, I believe that is expected since Opening the keyboard or change to MediaQuery would require UI rebuild Please see this answer for more details https://stackoverflow.com/a/65963713/8253662

**Comment by maheshmnj** [comment]
Closing as this is working as intended, if you disagree feel free to write back with a minimal code sample and the output of `flutter doctor -v`.

**Comment by b3nni97** [comment]
Yes, I am aware that it currently works as intended. However, that doesn't mean that you can't optimize the current solution.

I have put together a small example here which shows the problem:

https://gist.github.com/b3nni97/4bcdcd3872f5f6aecf11fab55d052e92

**1. click on + button
2. click on one of the text fields to open the keyboard
3. pay attention to the performance overlay
4. repeat step 1.**




I am testing it with an iPhone 12 Pro Max and iPhone X.
I have to open about 40 routes (2 times on + button) after that the app starts lagging as soon as you press on the text fields.

I know that normally this is rarely the case that you have so many routes open. However, as the pages get more complex, you need to open fewer routes to get the app to lag.

**I think there should be the possibility that if a route is not visible the media query is not applied (and thus no rebuild is generated).** 
Maybe there is currently a widget that forcibly blocks the subtree from being rebuilt, but I don't know about that.


[✓] Flutter (Channel master, 2.9.0-1.0.pre.220, on macOS 11.5.2 20G95 darwin-arm)

**Comment by maheshmnj** [comment]
Reopened based on the above comment.

**Comment by maheshmnj** [comment]
Thanks for the details @b3nni97,  this does look like a valid proposal. I will leave this issue open for further insights from the team


<details>
<summary>reproducible code sample</summary>

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

int GLOBAL_COUNTER = 0;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return CupertinoApp(
      showPerformanceOverlay: true,
      title: 'Flutter Demo',
      home: CupertinoTabScaffold(
        tabBar: CupertinoTabBar(items: const [
          BottomNavigationBarItem(
              icon: Icon(
                CupertinoIcons.home,
              ),
              label: "1"),
          BottomNavigationBarItem(
              icon: Icon(
                CupertinoIcons.home,
              ),
              label: "1"),
          BottomNavigationBarItem(
              icon: Icon(
                CupertinoIcons.home,
              ),
              label: "1"),
          BottomNavigationBarItem(
              icon: Icon(
                CupertinoIcons.home,
              ),
              label: "1"),
          BottomNavigationBarItem(
              icon: Icon(
                CupertinoIcons.home,
              ),
              label: "1"),
        ]),
        tabBuilder: (context, index) {
          return MyHomePage(
              title: 'Flutter Demo Home Page: ' + index.toString());
        },
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  void _incrementCounter() async {
    GLOBAL_COUNTER++;

    var route = CupertinoPageRoute(
      builder: (context) => MyHomePage(title: GLOBAL_COUNTER.toString()),
    );

    Navigator.of(context).push(route);

    if (GLOBAL_COUNTER % 20 != 0) {
      WidgetsBinding.instance!.addPostFrameCallback((_) {
        _incrementCounter();
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      navigationBar: CupertinoNavigationBar(
        middle: Text("Keyboard Page:" + GLOBAL_COUNTER.toString()),
      ),
      child: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const CupertinoTextField(),
            const CupertinoTextField(),
            const CupertinoTextField(),
            const CupertinoTextField(),
            const CupertinoTextField(),
            const CupertinoTextField(),
            const CupertinoTextField(),
            const CupertinoTextField(),
            const CupertinoTextField(),
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$GLOBAL_COUNTER',
              style: Theme.of(context).textTheme.headline4,
            ),
            FloatingActionButton(
              onPressed: _incrementCounter,
              tooltip: 'Increment',
              child: const Icon(Icons.add),
            ),
          ],
        ),
      ),
    );
  }
}

```
</details>

**Comment by b3nni97** [comment]
> Thanks for the details @b3nni97, this does look like a valid proposal. I will leave this issue open for further insights from the team.
> 
> > I know that normally this is rarely the case that you have so many routes open. However, as the pages get more complex, you need to open fewer routes to get the app to lag.
> 
> By the way can you please share your use case, where you would need to open so many routes and leave them in stack without navigating them?
> 
> minimal code sample

Yes, of course, the current use case is, for example:

- CupertinoTabScaffold exists. That means there are 4 pages loaded. 
- On page 1 - Navigate to the details of a product. (+1 Page)
- Click on Post (+1 Page)

So now in normal case 6 pages are opened.
These pages are very expensive i.e. they need approx. 9-12 ms in peak during a navigation in the CupertinoPageRoute in the build (this is in profile mode - however without DevTools it goes to 6-9 ms (says the performance overlay).

If you now want to write a comment on the post page, the keyboard opens and all pages are rebuilt, which of course causes the page the user is currently seeing to jank.

These individual pages can certainly be optimized a bit, but the MediaQuery usually rebuilds the CupertinoPageScaffold, i.e. all pages are rebuilt (and not just individual branches) and I don't know how to deny a subtree from being rebuilt.

**Comment by luckysmg** [comment]
I’ll take a look recently and file a PR to solve this issue O(∩_∩)O

**Comment by luckysmg** [comment]
See PR https://github.com/flutter/flutter/pull/116090

**Comment by otopba** [comment]
As temporary workaround, I use this code in my pages. This code prevents background pages from rebuild:
```dart
final currentRoute = ModalRoute.of(context)?.isCurrent ?? false;
if (!currentRoute) return const SizedBox.shrink();
```
